### DISCUSSION TITLE
Lesson 6, Part 18 - Cast Send & Call - retrieve() returns 0x

### DISCUSSION THREAD

**Discussion by 0xwtfwtf** [Cyfrin/foundry-full-course-cu#459] [General]
I used cast send in Foundry to "123" to the store function from SimpleStorage.sol in VSCode. I received output in the terminal that showed that the transaction went through - gas used, block number, etc. I pasted my terminal below so you can see.

But when I tried to call the retrieve function using cast call, it returns 0x. I tried Googling this but didn't find anything to fix it.

Here is my terminal - what did I do wrong?

Please ignore my first attempt to retrieve...I accidentally used curly quotes but you can see even when I fixed that, I still get 0x. I assume that means no data was stored so I must have done something wrong in the cast send even though a transaction went through.

So as an update, I was able to get the send and call functions to work when I deployed the contract directly (not using the script)...then I was able to retrieve 123 in hexadecimal. But I am still wondering why it returned 0x when I deployed using the script

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/138697105/e52f03a5-a04a-4f0f-9cf0-3d3b437d5c1f)

**Comment by shikhar229169** [comment]
I think the contract you are interacting with is not deployed with ```--broadcast```, so consider deploying the contract again with ```--broadcast``` at the end of command, and it would work.

So, when you don't add ```--broadcast```, only simulation happens on a temporary chain, it is actually not deployed.

**Reply by 0xwtfwtf** [reply]
I tried this and I am still getting 0x. Any other ideas?

**Reply by edoruk** [reply]
you should use different private key when you call 
```
cast send ${TX_ADDRESS} "store(uint256)" 123 --private-key ${DIFFERENT_PRIVATE_KEY}
```

**Reply by AbigirlMuchineripi** [reply]
@0xwtfwtf , kindly check the below screenshot for the contract address you need to use. It worked for me hope it will to you too.
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/94177156/02a497c2-f8a2-4fa5-8c82-7974e49349d9)

**Reply by Akasshhhh** [reply]
> I think the contract you are interacting with is not deployed with `--broadcast`, so consider deploying the contract again with `--broadcast` at the end of command, and it would work.
> 
> So, when you don't add `--broadcast`, only simulation happens on a temporary chain, it is actually not deployed.

Thanks brother, this worked.

**Reply by R0gue-one** [reply]
> @0xwtfwtf , kindly check the below screenshot for the contract address you need to use. It worked for me hope it will to you too. ![image](https://private-user-images.githubusercontent.com/94177156/328503044-02a497c2-f8a2-4fa5-8c82-7974e49349d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM2MTg3MDgsIm5iZiI6MTc0MzYxODQwOCwicGF0aCI6Ii85NDE3NzE1Ni8zMjg1MDMwNDQtMDJhNDk3YzItZjhhMi00ZmE1LThjODItNzk3NGU0OTM0OWQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAyVDE4MjY0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhZmMxOTkxMzVlYTY0NTBjOWU5NzlkNmMyNjIyZDllMDZiNTljMDg2YWE5MTJiNDQyOGExOGRhZDk3MjQ1MmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0VaXdqzHnJAVl4PF_iAeuVGuYcgiNEneTzrf9dgoD68)

This one fixed my problem I was thinking the [TO] was the address of function caller instead of contract address

**Reply by EngrPips** [reply]
Interesting.

**Comment by steinitz** [comment]
I continue to see `cast call` return `0x`.  I suspect others also see this.  We must be missing the secret sauce :) 

Here's my process:

1. create and compile a `SimpleStorage` contract as specified in the video, checked against the repo
2. run 'anvil' in the command line
3. set up `PRIVATE_KEY` and `RPC_URL` in `.env` based on the output of `anvil`
4. type `source .env` in the command line
5. copy an account from the output of anvil and use it in the following command line prompt:
`cast send 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 "store(uint256)" 123 --rpc-url $RPC_URL --private-key $PRIVATE_KEY`
6. examine `run-latest.json` to determine the contract address, as demonstrated by @AbigirlMuchineripi, above, and use it in:
`cast call 0x9fe46736679d2d9a65f0992f2272de9f3c7fa6e0 "retrieve()"`

which returns `0x`

**Reply by EngrPips** [reply]
I tried this myself, got my actual value, and then again later and was getting `0x.` I haven't revisited since then, but it is a nice time to revisit and see what I will find.

**Comment by steinitz** [comment]
Probably a red herring but the Foundry repo has an [issue](https://github.com/foundry-rs/foundry/issues/7764) about a failure when running `cast call`.  It's unclear whether the failure appeared in Apr or May.

In light of that, I ran `cast call` with the `--trace` option and saw no error.  But `--trace` changes the way `cast call` runs so it could sidestep the error.  The return value was unchanged: `'0x'`.

@EngrPips did you get a chance to revisit this?

**Reply by EngrPips** [reply]
Hmmm, I haven't revisited it but I will for sure drop whatever I found out when I do. Thanks for the heads-up man.

**Comment by PatrickAlphaC** [comment]
Wondering if you could shed light on this [issue](https://github.com/foundry-rs/foundry/issues/7764). I'm having a hard time getting ya'll error :/

**Reply by steinitz** [reply]
Yeah, that's the issue I referenced above.  In the end, I think it may be mostly a red herring because that issue actually generates an error, where we're seeing just an incorrect value.  Still, I think the incorrect value _is_ a Foundry-version issue.  The `cast send` and `cast call` seem to execute without issue so it would seem we've installed Foundry correctly.  Solid logic?

In any case, thanks for trying to reproduce our issue, @PatrickAlphaC.  Your stellar courses are much admired and your above-and-beyond follow-up support is much appreciated.

**Reply by EngrPips** [reply]
> Wondering if you could shed light on this [issue](https://github.com/foundry-rs/foundry/issues/7764). I'm having a hard time getting ya'll error :/

The issue is that we deploy our `SimpleStorage` contract to `Anvil` Blockchain, and then we used the `cast send` command to store a value on our `SimpleStorage` contract. After storing our value on the `SimpleStorage` contract, whenever we attempt to use `cast call` command to retrieve the value we stored on our `SimpleStorage` contract, we always get back `0x` as the value irrespective of whatever we stored in our contract which is not suppose to be the case

**Reply by steinitz** [reply]
I think @PatrickAlphaC was asking about the possibly-related [issue]( https://github.com/foundry-rs/foundry/issues/7764) in the Foundry repo.  Right?

**Reply by EngrPips** [reply]
Honestly, that issue is closely related to the issue we have here. The bottom line is `cast call` has been buggy since `2nd May build.`

**Reply by PatrickAlphaC** [reply]
If ya'll could add context to that issue in the foundry repo that would be great! It'll draw more attention to this issue that others will likely run into.

**Reply by EngrPips** [reply]
Yeah, we should!

**Reply by steinitz** [reply]
[Done](https://github.com/foundry-rs/foundry/issues/7764#issuecomment-2207504047)

**Reply by EngrPips** [reply]
Good

**Comment by steinitz** [comment]
I note the Foundry maintainers have made a fix and closed the [issue](https://github.com/foundry-rs/foundry/issues/7764) we reference above.   I did a quick google for how to update Foundry but came up short.  I'll have another look with a fresh mind in the next day or two and report back on whether it solves our `"0x"` issue.

I found the politics interesting: one developer said "Feel free to close this issue since this is technically not an issue with Foundry.".  But another stalwart developer went ahead with a subtle change which adds a helper function to put the result in "the `input` field _and_ the `data` field".  That change now satisfies two conflicting specifications: a newer one from ethereum.org and and older one from "most other provider docs".

Sounds encouraging to me.

**Reply by EngrPips** [reply]
Interesting, You can upgrade foundry by running `foundryup` in your terminal.

**Comment by steinitz** [comment]
The Foundry issue we've been following seems a red herring.  Foundry latest `cast call` still returns the wrong value when using @PatrickAlphaC's send/call process in his excellent video.  Maybe a recent Foundry change?

I found a [nice article](https://r4bbit.substack.com/p/smart-contract-development-with-foundry) which details the `create-send-call` process.  The "create" part, i.e. deployment with`forge create`, is necessary for me because, otherwise, I can't discover deployed contract address.  `Contract Address` is blank in the output of `cast send` and Foundry no longer creates a run-latest.json file, my previous source for the contract address, ever since I deleted the cache directory.  

Others may still have fresh run-latest.json so may not need the deployment step, below.

Here's a ten-step process, for Foundry latest, which results in a lovely, correct `0x00...07b` result

1. type in the terminal command line  
`foundryup` 
to get the latest nightly build of forge, anvil, cast... (Thanks @EngrPips)
2. create and compile a `SimpleStorage` contract as specified in the video, checked against the repo
3. type in the terminal command line  
`anvil`
4. set up `PRIVATE_KEY` and `RPC_URL` in `.env` based on the output of `anvil` (they never seem to change)
5. leave the `anvil` terminal and launch a second terminal
6. type in the terminal command line 
`source .env` 
7. type in the terminal command line
`forge create --rpc-url $RPC_URL --private-key $PRIVATE_KEY src/SimpleStorage.sol:SimpleStorage`
8. take note of `Deployer` and `Deploy To` from the `forge create` output.  `Deploy To` is the contract address
9. replace  \<Deploy To\> and \<Deployer\>, below, with the actual values from the `forge create` output, and run it in the terminal command line:
`cast send <Deploy To> "store(uint256)" 123 --rpc-url $RPC_URL --private-key $PRIVATE_KEY --from <Deployer>`
10. finally, replace  \<Deploy To\>, below, with the value from `forge create`, and run it in the terminal command line:
`cast call <Deploy To> "retrieve()"`
11. rejoice at the expected result: `0x00...07b`

**Reply by EngrPips** [reply]
Thanks a bunch for taking the time to explain in detail this process. We are grateful.

**Reply by tiznah** [reply]
Thank you!

**Comment by yangpili1997** [comment]
cast send to 后面跟的是启动anvil的Contract Address：xxxx

**Reply by EngrPips** [reply]
what does this `后面跟的是启动anvil的` mean?
